oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java

```diff
@@ -610,42 +615,79 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
-            headerBuf.writeInt(sequenceId);
-            headerBuf.writeBoolean(hasNextBatch);
-            headerBuf.writeInt(columnCount);
-            headerBuf.writeInt(rowCount);
-            for (int dataBufferLength : dataBufferLengths) {
-                headerBuf.writeInt(dataBufferLength);
-            }
-            for (boolean isComposited : compositeFlags) {
-                headerBuf.writeBoolean(isComposited);
-            }
-            headerBuf.writeInt(compositedElementCount);
-
-            // write to channel
-            out.write(headerBuf);
-            out.write(compositedElementLengths);
-            for (int i = 0; i < columnCount; i++) {
-                out.write(dataBuffers[i]);
-            }
-            if (columnCount > 0) {
+            originLength += (columnCount * rowCount); // for null buffers
+            if (!compressEnabled) {
+                // set header
+                ByteBuf headerBuf = allocateBuffer(allocator, ID, headerLength,
+                        originLength, false);
+                headerBuf.writeInt(sequenceId);
+                headerBuf.writeBoolean(hasNextBatch);
+                headerBuf.writeInt(columnCount);
+                headerBuf.writeInt(rowCount);
+                for (int dataBufferLength : dataBufferLengths) {
+                    headerBuf.writeInt(dataBufferLength);
+                }
+                for (boolean isComposited : compositeFlags) {
+                    headerBuf.writeBoolean(isComposited);
+                }
+                headerBuf.writeInt(compositedElementCount);
+                // track content length which will be used for decompress
+                headerBuf.writeBoolean(false);
+                headerBuf.writeInt(originLength);
+                out.write(headerBuf);
+                // write content to channel
+                out.write(compositedElementLengths);
+                for (int i = 0; i < columnCount; i++) {
+                    out.write(dataBuffers[i]);
+                }
                 int columnIndex = 0;
                 for (; columnIndex < columnCount - 1; columnIndex++) {
                     out.write(nullBuffers[columnIndex]);
                 }
                 out.write(nullBuffers[columnIndex], promise);
+            } else {
+                // compress data first to calculate contentLength
+                CompositeByteBuf compositeByteBuf = Unpooled.compositeBuffer(2 * columnCount + 1);
+                compositeByteBuf.addComponent(compositedElementLengths);
+                for (int i = 0; i < columnCount; i++) {
+                    compositeByteBuf.addComponent(dataBuffers[i]);
+                }
+                for (int i = 0; i < columnCount; i++) {
+                    compositeByteBuf.addComponent(nullBuffers[i]);
+                }
+                try {
+                    ByteBuf compressedData = ZStdUtils.compress(compositeByteBuf);

```

Comment:
It's ok to directly use `ZStdUtils` here. But in the future we may support more kind of codec, user could set codec type in profile. Let's do a refactor in later PRs.

Reply:
Yes. sure.